File: wallet/packages/blockchain/src/custody.ts
/**
 * Custody tiers, defined by what it takes to move money out of them
 * (ADR-0018, prompt_phase4.md rules 134-139).
 *
 * THE TRAP THIS MODULE EXISTS TO AVOID
 *
 * Rule 137: tiers that all sign the same way are labels, not segregation. An
 * enum that three code paths ignore gives an auditor, an operator and a reader
 * of the schema the appearance of custody separation while providing none —
 * which is worse than having no tiers at all, because it is load-bearing in
 * everyone's mental model and nowhere else.
 *
 * So a tier is not a name. A tier is a POLICY: which authorities must have
 * signed the authorization proof before key material will produce a signature.
 *
 * WHERE THIS IS ENFORCED
 *
 * The policy is declared here, in chain-agnostic terms, so the domain can
 * reason about it. It is ENFORCED inside the signing boundary, by participants
 * that verify the proof against a key they hold independently of the
 * coordinator (ADR-0015). A tier check that lived only in the API would be a
 * check a compromised API could skip, and the API is exactly the component the
 * threshold scheme assumes may be compromised.
 */

/** Ordered from most to least exposed. */
export const CUSTODY_TIERS = ['deposit', 'hot', 'warm', 'cold'] as const;

export type CustodyTier = (typeof CUSTODY_TIERS)[number];

/**
 * Who must have signed an authorization for a tier to move.
 *
 * `risk-engine` is the automated policy engine. `operator` is a human with an
 * authenticated identity. `ceremony` is an out-of-band procedure with more than
 * one human, recorded in a runbook.
 */
export type Authority = 'risk-engine' | 'operator' | 'ceremony';

export interface TierPolicy {
  readonly tier: CustodyTier;
  /** Every authority in this list must appear in the proof. */
  readonly requiredAuthorities: readonly Authority[];
  /** Signing participants required, where the tier uses threshold signing. */
  readonly signingThreshold: number;
  /**
   * True when the tier's key may only ever pay one hardcoded destination
   * (ADR-0005, ADR-0017). The deposit class is the reason that ADR was
   * acceptable, and honouring it here is what makes it true rather than a
   * story we told ourselves.
   */
  readonly fixedDestinationOnly: boolean;
  readonly online: boolean;
}

/**
 * The asymmetry is deliberate: moving money INTO colder storage is automatic,
 * moving it OUT is the privileged operation. Getting the conservative direction
 * wrong costs an unnecessary transfer; getting the other one wrong costs the
 * treasury.
 */
export const TIER_POLICIES: Readonly<Record<CustodyTier, TierPolicy>> = {
  deposit: {
    tier: 'deposit',
    // No authority list: a deposit key cannot choose a destination, so there is
    // nothing for an authorization to authorise beyond "sweep now".
    requiredAuthorities: [],
    signingThreshold: 1,
    fixedDestinationOnly: true,
    online: true,
  },
  hot: {
    tier: 'hot',
    requiredAuthorities: ['risk-engine'],
    signingThreshold: 3,
    fixedDestinationOnly: false,
    online: true,
  },
  warm: {
    tier: 'warm',
    requiredAuthorities: ['risk-engine', 'operator'],
    signingThreshold: 3,
    fixedDestinationOnly: false,
    online: true,
  },
  cold: {
    tier: 'cold',
    // Two humans, out of band. No automated authority can satisfy this.
    requiredAuthorities: ['ceremony'],
    signingThreshold: 4,
    fixedDestinationOnly: false,
    online: false,
  },
};

/**
 * Parse the authorities a proof's `approvedBy` claims.
 *
 * The format is `authority[:identity]`, joined by `+`:
 *
 *   risk-engine
 *   risk-engine+operator:alice
 *   ceremony:2026-09-11
 *
 * An unrecognised authority is DROPPED rather than accepted, so a proof
 * claiming `superuser` satisfies nothing. Failing open here would make the
 * whole policy decorative.
 */
export function parseAuthorities(approvedBy: string): readonly Authority[] {
  const known = new Set<string>(['risk-engine', 'operator', 'ceremony']);
  return approvedBy
    .split('+')
    .map((part) => part.split(':')[0]?.trim() ?? '')
    .filter((name): name is Authority => known.has(name));
}

export type TierCheck =
  | { readonly ok: true }
  | { readonly ok: false; readonly missing: readonly Authority[] };

/**
 * Does this proof carry everything the tier requires?
 *
 * Deliberately NOT a risk evaluation. It checks the ATTESTATION, not the
 * policy — re-running risk rules here would put policy in two places and
 * violate master-prompt rule 109, which keeps authorising and signing separate.
 */
/**
 * Does `held` satisfy a requirement for `required`?
 *
 * THE ASYMMETRY THAT MATTERS
 *
 * A human operator satisfies a requirement for the risk engine: a named person
 * deliberately approving a movement is strictly more authority than automation
 * applying a policy, and a `ceremony` is more than either. The reverse is never
 * true — automation must never satisfy a requirement for a human, because that
 * is precisely the substitution a compromised coordinator would like to make.
 *
 * Without this, an operator acting directly on the hot tier is refused for
 * carrying too much authority rather than too little, which is nonsense — and
 * the first thing anyone hitting it would reach for is a proof falsely claiming
 * `risk-engine`, defeating the point of naming the authority at all.
 */
function satisfies(held: Authority, required: Authority): boolean {
  if (held === required) return true;
  if (required === 'risk-engine') return held === 'operator' || held === 'ceremony';
  // An `operator` requirement is NOT satisfied by a ceremony: they are
  // different procedures, and a ceremony proof names no individual.
  return false;
}

/**
 * Every required authority must be met by a DISTINCT authority in the proof.
 *
 * This is the correction that matters, and it was nearly missed. Warm requires
 * `[risk-engine, operator]` — meaning **two independent approvals**, not "an
 * authority level of at least operator". With a naive check, a lone `operator`
 * satisfies the operator requirement AND, through the hierarchy above, the
 * risk-engine one — silently reducing warm from two approvals to one, which is
 * the entire protection warm exists to provide.
 *
 * Requirements are matched most-constrained first (an `operator` requirement
 * has one possible satisfier, a `risk-engine` requirement has three), which is
 * exact for this nested structure and needs no backtracking.
 */
function countSatisfiers(required: Authority): number {
  return (['risk-engine', 'operator', 'ceremony'] as const).filter((held) =>
    satisfies(held, required),
  ).length;
}

export function checkTierAuthorization(tier: CustodyTier, approvedBy: string): TierCheck {
  const policy = TIER_POLICIES[tier];
  const available = parseAuthorities(approvedBy).slice();

  const ordered = [...policy.requiredAuthorities].sort(
    (a, b) => countSatisfiers(a) - countSatisfiers(b),
  );

  const missing: Authority[] = [];
  for (const required of ordered) {
    const index = available.findIndex((held) => satisfies(held, required));
    if (index === -1) missing.push(required);
    // Consumed, so one approval cannot count twice.
    else available.splice(index, 1);
  }

  return missing.length === 0 ? { ok: true } : { ok: false, missing };
}

/**
 * Which tier should fund a withdrawal of this size.
 *
 * Ordinary withdrawals come from hot. One that hot cannot cover is an
 * OPERATIONAL EVENT, not an error: it means the float needs topping up, and
 * that is a decision with a human in it (rule 135).
 */
export function tierForWithdrawal(input: {
  readonly amount: bigint;
  readonly hotBalance: bigint;
}): CustodyTier | 'insufficient_hot' {
  return input.amount <= input.hotBalance ? 'hot' : 'insufficient_hot';
}

export interface RebalancePlan {
  readonly action: 'none' | 'to_warm' | 'from_warm';
  readonly amount: bigint;
}

/**
 * Threshold-based rebalancing (rule 135).
 *
 * Above the ceiling, the excess goes to warm — automatic, because it moves
 * value to safety. Below the floor, a top-up is REQUESTED rather than
 * performed, because it moves value out of warm and that requires an operator
 * proof (ADR-0018).
 */
export function planRebalance(input: {
  readonly hotBalance: bigint;
  readonly ceiling: bigint;
  readonly floor: bigint;
  readonly target: bigint;
}): RebalancePlan {
  if (input.ceiling < input.floor) {
    throw new Error('hot ceiling must not be below the floor');
  }
  if (input.hotBalance > input.ceiling) {
    return { action: 'to_warm', amount: input.hotBalance - input.target };
  }
  if (input.hotBalance < input.floor) {
    return { action: 'from_warm', amount: input.target - input.hotBalance };
  }
  return { action: 'none', amount: 0n };
}

File: wallet/services/mpc/src/custody.rs
//! Custody tiers, enforced where a compromised coordinator cannot skip them.
//!
//! ADR-0018, prompt_phase4.md rules 134-139.
//!
//! # Why this lives in the signing service
//!
//! The TypeScript side has the same policy table (`packages/blockchain/
//! src/custody.ts`) and the domain reasons about tiers with it. That copy is
//! advisory. **This one decides.**
//!
//! ADR-0015 establishes that the API is the coordinator and that a compromised
//! coordinator must be a liveness problem rather than a safety one. A tier
//! check that lived only in the API would be a check the compromised component
//! performs on itself. So the authority requirements are re-derived here, from
//! the tier named in the request, against a proof signed by a key this service
//! holds independently of the caller.
//!
//! Rule 137 is the thing being defended: tiers that all sign the same way are
//! labels, not segregation.

use serde::{Deserialize, Serialize};

/// Ordered from most to least exposed.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum CustodyTier {
    Deposit,
    Hot,
    Warm,
    Cold,
}

/// Who must have signed an authorisation for a tier to move.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Authority {
    RiskEngine,
    Operator,
    Ceremony,
}

impl Authority {
    fn parse(name: &str) -> Option<Self> {
        match name.trim() {
            "risk-engine" => Some(Self::RiskEngine),
            "operator" => Some(Self::Operator),
            "ceremony" => Some(Self::Ceremony),
            // Anything else satisfies NOTHING. Failing open here would make
            // the policy decorative: a proof claiming `superuser` must not
            // count as an operator.
            _ => None,
        }
    }

    pub fn as_str(self) -> &'static str {
        match self {
            Self::RiskEngine => "risk-engine",
            Self::Operator => "operator",
            Self::Ceremony => "ceremony",
        }
    }
}

impl CustodyTier {
    /// Every authority that must appear in the proof for this tier.
    ///
    /// The asymmetry is deliberate: moving value INTO colder storage is
    /// automatic, moving it OUT is privileged. Getting the conservative
    /// direction wrong costs a transfer; getting the other one wrong costs the
    /// treasury.
    pub fn required_authorities(self) -> &'static [Authority] {
        match self {
            // A deposit key cannot choose a destination (ADR-0005), so there
            // is nothing for an authorisation to authorise beyond "sweep now".
            Self::Deposit => &[],
            Self::Hot => &[Authority::RiskEngine],
            Self::Warm => &[Authority::RiskEngine, Authority::Operator],
            // Two humans, out of band. No automated authority satisfies this.
            Self::Cold => &[Authority::Ceremony],
        }
    }

    /// Signing participants required. Secondary to the authority check: raising
    /// the threshold defends against participant compromise, while the
    /// authority requirement defends against a compromised coordinator.
    pub fn signing_threshold(self) -> u8 {
        match self {
            Self::Deposit => 1,
            Self::Hot | Self::Warm => 3,
            Self::Cold => 4,
        }
    }

    /// True when this key class may only ever pay one hardcoded destination.
    pub fn fixed_destination_only(self) -> bool {
        matches!(self, Self::Deposit)
    }
}

/// Parse the authorities an `approved_by` field claims.
///
/// Format is `authority[:identity]`, joined by `+`:
///
/// ```text
/// risk-engine
/// risk-engine+operator:alice
/// ceremony:2026-09-11
/// ```
pub fn parse_authorities(approved_by: &str) -> Vec<Authority> {
    approved_by
        .split('+')
        .filter_map(|part| Authority::parse(part.split(':').next().unwrap_or("")))
        .collect()
}

/// Does this proof carry everything the tier requires?
///
/// Returns the first missing authority, for an error message an operator can
/// act on. Deliberately NOT a risk evaluation: this checks the attestation,
/// not the policy (master-prompt rule 109).
/// Does `held` satisfy a requirement for `required`?
///
/// # The asymmetry that matters
///
/// A human operator satisfies a requirement for the risk engine: a named person
/// deliberately approving a movement is strictly more authority than automation
/// applying a policy, and a ceremony is more than either. The reverse is never
/// true — automation must never satisfy a requirement for a human, because that
/// is exactly the substitution a compromised coordinator would like to make.
fn satisfies(held: Authority, required: Authority) -> bool {
    if held == required {
        return true;
    }
    match required {
        Authority::RiskEngine => matches!(held, Authority::Operator | Authority::Ceremony),
        // An operator requirement is NOT satisfied by a ceremony: they are
        // different procedures, and a ceremony proof names no individual.
        _ => false,
    }
}

/// How many authorities could satisfy this requirement. Used to match the
/// most-constrained requirement first.
fn satisfier_count(required: Authority) -> usize {
    [
        Authority::RiskEngine,
        Authority::Operator,
        Authority::Ceremony,
    ]
    .iter()
    .filter(|held| satisfies(**held, required))
    .count()
}

/// Does this proof carry everything the tier requires?
///
/// Every required authority must be met by a **distinct** authority in the
/// proof. Warm requires `[risk-engine, operator]`, meaning two independent
/// approvals — not "an authority level of at least operator". Without the
/// distinctness rule a lone operator satisfies the operator requirement and,
/// through the hierarchy above, the risk-engine one as well, silently reducing
/// warm from two approvals to one: the entire protection warm exists to give.
///
/// Requirements are matched most-constrained first, which is exact for this
/// nested structure and needs no backtracking.
///
/// Returns the first unmet authority, for an error an operator can act on.
/// Deliberately NOT a risk evaluation: this checks the attestation, not the
/// policy (master-prompt rule 109).
pub fn check_tier_authorization(tier: CustodyTier, approved_by: &str) -> Option<Authority> {
    let mut available = parse_authorities(approved_by);

    let mut ordered: Vec<Authority> = tier.required_authorities().to_vec();
    ordered.sort_by_key(|required| satisfier_count(*required));

    for required in ordered {
        match available.iter().position(|held| satisfies(*held, required)) {
            // Consumed, so one approval cannot count twice.
            Some(index) => {
                available.remove(index);
            }
            None => return Some(required),
        }
    }

    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn hot_accepts_the_risk_engine_alone() {
        assert!(check_tier_authorization(CustodyTier::Hot, "risk-engine").is_none());
    }

    #[test]
    fn warm_refuses_automation_alone() {
        // The property that makes the tier real: a compromised API can produce
        // a risk-engine proof, but not an operator's signature.
        assert_eq!(
            check_tier_authorization(CustodyTier::Warm, "risk-engine"),
            Some(Authority::Operator)
        );
    }

    #[test]
    fn warm_accepts_both() {
        assert!(
            check_tier_authorization(CustodyTier::Warm, "risk-engine+operator:alice").is_none()
        );
    }

    #[test]
    fn cold_refuses_everything_automated() {
        assert!(
            check_tier_authorization(CustodyTier::Cold, "risk-engine+operator:alice").is_some()
        );
        assert!(check_tier_authorization(CustodyTier::Cold, "ceremony:2026-09-11").is_none());
    }

    #[test]
    fn unknown_authorities_satisfy_nothing() {
        assert!(parse_authorities("superuser+root").is_empty());
        assert!(check_tier_authorization(CustodyTier::Warm, "superuser").is_some());
    }

    #[test]
    fn an_authority_named_as_an_identity_does_not_count() {
        // `operator:risk-engine` names ONE authority, not two.
        assert_eq!(
            parse_authorities("operator:risk-engine"),
            vec![Authority::Operator]
        );
        assert!(check_tier_authorization(CustodyTier::Warm, "operator:risk-engine").is_some());
    }

    #[test]
    fn an_operator_satisfies_a_hot_requirement_alone() {
        // Surfaced by the nonce-provisioning script: an operator acting
        // directly on hot was refused for carrying too MUCH authority, and the
        // obvious workaround would have been a proof falsely claiming
        // `risk-engine`.
        assert!(check_tier_authorization(CustodyTier::Hot, "operator:alice").is_none());
        assert!(check_tier_authorization(CustodyTier::Hot, "ceremony:2026-09-11").is_none());
    }

    #[test]
    fn automation_never_satisfies_a_human_requirement() {
        // The direction that matters: exactly the substitution a compromised
        // coordinator would like to make.
        assert!(check_tier_authorization(CustodyTier::Warm, "risk-engine").is_some());
        assert!(check_tier_authorization(CustodyTier::Cold, "risk-engine").is_some());
    }

    #[test]
    fn a_ceremony_does_not_stand_in_for_a_named_operator() {
        assert_eq!(
            check_tier_authorization(CustodyTier::Warm, "ceremony:2026-09-11"),
            Some(Authority::Operator)
        );
    }

    #[test]
    fn warm_means_two_approvals_not_one_senior_one() {
        // The subtle one. With a naive hierarchy check, `operator` satisfies
        // the operator requirement AND the risk-engine requirement, silently
        // reducing warm from two independent approvals to one — the whole
        // protection warm exists to give.
        assert!(check_tier_authorization(CustodyTier::Warm, "operator:alice").is_some());
        assert!(
            check_tier_authorization(CustodyTier::Warm, "risk-engine+operator:alice").is_none()
        );
        assert!(
            check_tier_authorization(CustodyTier::Warm, "ceremony:2026-09-11+operator:alice")
                .is_none()
        );
        // Hot needs only one, so a single operator is enough there.
        assert!(check_tier_authorization(CustodyTier::Hot, "operator:alice").is_none());
    }

    #[test]
    fn no_two_tiers_share_a_policy() {
        // The executable form of rule 137. Tiers that all sign the same way
        // are labels, and this fails the moment that becomes true.
        let tiers = [
            CustodyTier::Deposit,
            CustodyTier::Hot,
            CustodyTier::Warm,
            CustodyTier::Cold,
        ];
        let mut signatures: Vec<String> = tiers
            .iter()
            .map(|t| {
                let mut names: Vec<&str> = t
                    .required_authorities()
                    .iter()
                    .map(|a| a.as_str())
                    .collect();
                names.sort_unstable();
                format!(
                    "{}|{}|{}",
                    names.join("+"),
                    t.signing_threshold(),
                    t.fixed_destination_only()
                )
            })
            .collect();
        signatures.sort();
        signatures.dedup();
        assert_eq!(signatures.len(), tiers.len());
    }

    #[test]
    fn only_deposit_keys_are_destination_locked() {
        assert!(CustodyTier::Deposit.fixed_destination_only());
        for tier in [CustodyTier::Hot, CustodyTier::Warm, CustodyTier::Cold] {
            assert!(!tier.fixed_destination_only());
        }
    }

    #[test]
    fn deposit_requires_no_authority_because_it_cannot_choose_a_destination() {
        assert!(check_tier_authorization(CustodyTier::Deposit, "").is_none());
    }
}
